File: doc/jvsipDataModel/codeEx/strassFunc.c
vsip_mview_d * VU_strass_d(vsip_mview_d *A, vsip_mview_d *B, vsip_mview_d *C){
   vsip_length n=vsip_mgetrowlength_d(A);
   if((n % 2) | (n < 129)){
      vsip_mprod_d(A,B,C);
   } else {
      vsip_mview_d *A00=vsip_mcloneview_d(A); vsip_mview_d *Aij=vsip_mcloneview_d(A);
      vsip_mview_d *Ai0=vsip_mcloneview_d(A); vsip_mview_d *A0j=vsip_mcloneview_d(A);
      vsip_mview_d *B00=vsip_mcloneview_d(B); vsip_mview_d *Bij=vsip_mcloneview_d(B);
      vsip_mview_d *Bi0=vsip_mcloneview_d(B); vsip_mview_d *B0j=vsip_mcloneview_d(B);
      vsip_mview_d *C00=vsip_mcloneview_d(C); vsip_mview_d *Cij=vsip_mcloneview_d(C);
      vsip_mview_d *Ci0=vsip_mcloneview_d(C); vsip_mview_d *C0j=vsip_mcloneview_d(C);
      vsip_length m=n/2;
      vsip_index i=(vsip_index) m, j=i;
      vsip_mview_d *P1=vsip_mcreate_d(m,m,VSIP_ROW,VSIP_MEM_NONE);
      vsip_mview_d *P2=vsip_mcreate_d(m,m,VSIP_ROW,VSIP_MEM_NONE);
      vsip_mview_d *P3=vsip_mcreate_d(m,m,VSIP_ROW,VSIP_MEM_NONE);
      vsip_mview_d *P4=vsip_mcreate_d(m,m,VSIP_ROW,VSIP_MEM_NONE);
      vsip_mview_d *P5=vsip_mcreate_d(m,m,VSIP_ROW,VSIP_MEM_NONE);
      vsip_mview_d *P6=vsip_mcreate_d(m,m,VSIP_ROW,VSIP_MEM_NONE);
      vsip_mview_d *P7=vsip_mcreate_d(m,m,VSIP_ROW,VSIP_MEM_NONE);
      VU_MGETSUB_D(A,A00,0,0,1,1,m,m);VU_MGETSUB_D(A,A0j,0,j,1,1,m,m);
      VU_MGETSUB_D(A,Ai0,i,0,1,1,m,m);VU_MGETSUB_D(A,Aij,i,j,1,1,m,m);
      VU_MGETSUB_D(B,B00,0,0,1,1,m,m);VU_MGETSUB_D(B,B0j,0,j,1,1,m,m);
      VU_MGETSUB_D(B,Bi0,i,0,1,1,m,m);VU_MGETSUB_D(B,Bij,i,j,1,1,m,m);
      VU_MGETSUB_D(C,C00,0,0,1,1,m,m);VU_MGETSUB_D(C,C0j,0,j,1,1,m,m);
      VU_MGETSUB_D(C,Ci0,i,0,1,1,m,m);VU_MGETSUB_D(C,Cij,i,j,1,1,m,m);
      vsip_madd_d(A00,Aij,C00); vsip_madd_d(B00,Bij,Cij); VU_strass_d(C00,Cij,P1);
      vsip_madd_d(Ai0,Aij,C00); VU_strass_d(C00,B00,P2); 
      vsip_msub_d(B0j,Bij,C00); VU_strass_d(A00,C00,P3);
      vsip_msub_d(Bi0,Bij,C00); VU_strass_d(Aij,C00,P4); 
      vsip_madd_d(A00,A0j,C00); VU_strass_d(C00,Bij,P5);
      vsip_msub_d(Ai0,A00,C00); vsip_madd_d(B00,B0j,Cij); VU_strass_d(C00,Cij,P6);
      vsip_msub_d(A0j,Aij,C00); vsip_madd_d(Bi0,Bij,Cij); VU_strass_d(C00,Cij,P7);
      vsip_madd_d(P1,P4,C00);vsip_msub_d(C00,P5,C00); 
      vsip_madd_d(C00,P7,C00); vsip_madd_d(P3,P5,C0j); 
      vsip_madd_d(P2,P4,Ci0); vsip_madd_d(P1,P3,Cij); 
      vsip_msub_d(Cij,P2,Cij); vsip_madd_d(Cij,P6,Cij);
      vsip_mdestroy_d(C00); vsip_mdestroy_d(Cij); vsip_mdestroy_d(Ci0);
      vsip_mdestroy_d(C0j); vsip_mdestroy_d(A00); vsip_mdestroy_d(Ai0); 
      vsip_mdestroy_d(A0j);vsip_mdestroy_d(Aij); vsip_mdestroy_d(B00);
      vsip_mdestroy_d(Bi0); vsip_mdestroy_d(B0j);vsip_mdestroy_d(Bij);
   }
   return C;

File: doc/jvsipDataModel/codeEx/strassFunc.py
def strass(A, B, C):
    assert 'View' in repr(A) and 'View' in repr(B) and 'View' in repr(C),\
          'Strass works with pyJvsip views'
    assert 'mview' in A.type,'strass only works with matrix views'
    assert A.type==B.type and A.type==C.type,'Matrix types must be the same'
    n=A.rowlength
    if n % 2 == 1  or n < 129:
        pjv.prod(A,B,C)
    else:
        A00 = A.cloneview; Aij = A.cloneview
        Ai0 = A.cloneview; A0j = A.cloneview
        B00 = B.cloneview; Bij = B.cloneview
        Bi0 = B.cloneview; B0j = B.cloneview
        C00 = C.cloneview; Cij = C.cloneview
        Ci0 = C.cloneview; C0j = C.cloneview
        m = n/2; i = m; j = i
        P1 = pjv.create(A.type,m,m)
        P2 = pjv.create(A.type,m,m)
        P3 = pjv.create(A.type,m,m)
        P4 = pjv.create(A.type,m,m)
        P5 = pjv.create(A.type,m,m)
        P6 = pjv.create(A.type,m,m)
        P7 = pjv.create(A.type,m,m)
        mgetsub(A,A00,0,0,1,1,m,m);mgetsub(A,A0j,0,j,1,1,m,m)
        mgetsub(A,Ai0,i,0,1,1,m,m);mgetsub(A,Aij,i,j,1,1,m,m)
        mgetsub(B,B00,0,0,1,1,m,m);mgetsub(B,B0j,0,j,1,1,m,m)
        mgetsub(B,Bi0,i,0,1,1,m,m);mgetsub(B,Bij,i,j,1,1,m,m)
        mgetsub(C,C00,0,0,1,1,m,m);mgetsub(C,C0j,0,j,1,1,m,m)
        mgetsub(C,Ci0,i,0,1,1,m,m);mgetsub(C,Cij,i,j,1,1,m,m)
        pjv.add(A00,Aij,C00); pjv.add(B00,Bij,Cij); strass(C00,Cij,P1)
        pjv.add(Ai0,Aij,C00);strass(C00,B00,P2)
        pjv.sub(B0j,Bij,C00); strass(A00,C00,P3)
        pjv.sub(Bi0,Bij,C00); strass(Aij,C00,P4)
        pjv.add(A00,A0j,C00); strass(C00,Bij,P5)
        pjv.sub(Ai0,A00,C00); pjv.add(B00,B0j,Cij); strass(C00,Cij,P6)
        pjv.sub(A0j,Aij,C00); pjv.add(Bi0,Bij,Cij); strass(C00,Cij,P7)
        pjv.add(P1,P4,C00);pjv.sub(C00,P5,C00); pjv.add(C00,P7,C00)
        pjv.add(P3,P5,C0j)
        pjv.add(P2,P4,Ci0)
        pjv.add(P1,P3,Cij); pjv.sub(Cij,P2,Cij); pjv.add(Cij,P6,Cij)
    return C

